### ISSUE TITLE
Multiple Ctrl+C is needed to finish every process and thread in the early phase

### ISSUE THREAD

**Issue by YuriCat** [DeNA/HandyRL#269]
It has been a little bothering me for a long time, but most of the time when I try to stop it with Ctrl+C early in the training process, it doesn't stop at once.

It seems that it's stopping at the following `Queue.get()`:

```
^CTraceback (most recent call last):
  File "/Users/ohto/myrepo/HandyRL/handyrl/train.py", line 644, in run
    self.server()
  File "/Users/ohto/myrepo/HandyRL/handyrl/train.py", line 568, in server
    conn, (req, data) = self.worker.recv()
  File "/Users/ohto/myrepo/HandyRL/handyrl/connection.py", line 221, in recv
    return self.input_queue.get()
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/queue.py", line 171, in get
    self.not_empty.wait()
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 312, in wait
    waiter.acquire()
KeyboardInterrupt
```

However, setting `timeout` did not affect this.
https://github.com/DeNA/HandyRL/compare/develop...YuriCat:experiment/ctrl_c_queue_timeout

**Comment by baichii** [comment]
I had the same problem. Subprocess can‘t stop while master process finished. 
The main reason is that conn.recv() will wait forever in the  subprocess, so self.shutdown_flag can't work, after fix this problem, I found the subprocess will copy(or like this)  a method from the master process to its own, so the param in the master process has no influence in subprocess(like self.shutdown_flag)  any more，change it  in the master process but it did nothing to subprocess(sad).

Add a timeout check may save it  but it's not a good idea.I turn to single process.

```
    def _worker(self, conn, bid):
        print('started batcher %d' % bid)
        while not self.shutdown_flag:
            # original version 
            # episodes = conn.recv()
            # batch = cal(episodes)
            # conn.send(batch)
            while conn.poll(timeout=0.8):
                episodes = conn.recv()
                batch = cal(episodes)
                conn.send(batch)
            else:
                break
        print('finished batcher %d' % bid)
```

**Comment by YuriCat** [comment]
@baichii  Thank you for your report and great idea!

We are currently trying this PR #307.
In this PR, all threads are started with `daemon=True` and then easily shutdown after the first Ctrl+C.
In addition, we can also finish every process after training until configured 'epochs' in config.yaml.

As you pointed out, how to stop processes inside the Batcher is also an important issue.
We have set `daemon=True` in this PR, but using `multiprocessing.Value` is also a great idea to be considered.

**Comment by baichii** [comment]
Cool, set daemon=True  is a good idea, and i'll use this trick sooner or later. And i found conn.send() may get blocked too, only check timeout before conn.recv() is not always effective.

**Comment by baichii** [comment]
Cool, use daemon is a good idea, I'll have a try.
And I find another problem in Batcher is that timeout occurs not only at conn.recv() but also at conn.send(). No parent conn recv the data from a child conn so it blocked. 
I add some recv in shutdown() to make sure child conn could send all data and close.  After this, all processes seem  work well and could close automatically.
```
class MultiProcessJobExecutor:

    def shutdown(self):
        self.shutdown_flag = True 
        # make sure conn closed
        for conn in self.conns:
            conn.recv()
        for thread in self.threads:
            thread.join()

**Comment by YuriCat** [comment]
Merged #307.
